### ISSUE TITLE
Poor multi threaded performance

### ISSUE THREAD

**Issue by gauravkumar37** [arangodb/arangodb-java-driver#75]
I am using the latest driver v4.1.2 in Scala with latest ArangoDB v3.1.1
When I am using a single thread to get a document by its key, time taken is usually 0-2 ms. But when I am using multiple threads to concurrently get documents from the DB, performance degrades and time measured is around 30 ms.
I dug into the code and did some profiling and found out the following possible bottlenecks:

1. Every time it tries to write something to the connection, it first checks whether the connection is open or not. Checking this is currently synchronized leading to thread locks. (Connection.java:74)
2. Afterwards while writing the message to connection, it is again synchronized (ConnectionSync.java:88)

Profiling the code gave me similar results- CPU time spent is minimal but total time spent is a lot.

![arangodbperftest](https://cloud.githubusercontent.com/assets/2500570/20454761/11061252-ae70-11e6-848a-efc6931ae7ee.PNG)

Can you also please share the code used in this blog: https://www.arangodb.com/2016/10/updated-java-drivers-with-arangodb-3-1/

**Comment by mpv1989** [comment]
Hi,

we tested the performance with the Yahoo! Cloud System Benchmark (YCSB).
You can find our code in my fork [https://github.com/mpv1989/YCSB](https://github.com/mpv1989/YCSB).

The synchronized methods in `Connection`, `ConnectionSync` are maybe not all necessary, I will check this.

For the best performance every thread should use its own instance of `ArangoDB`. So every thread creates its own connection to the server.

Best
Mark

**Comment by gauravkumar37** [comment]
Thanks @mpv1989, will definitely check the code out.
I tested with both configurations- one driver instance per JVM and separate driver instances per thread.

My results so far:
In my cluster v3.1.1, with driver v3.1 I achieved a max of 6500 sync per second with 40 threads and 1 driver with 40 HTTP connections. With HTTP version of the driver v3.1 I only need 1 driver instance since it can open multiple HTTP connections to server.
With new driver v4.1 sync, with 40 threads each having a separate driver instance of ArangoDB, I could only achieve a max of 4000.

What could be the possible reasons for the same?
